### PR TITLE
Support for api_url

### PR DESCRIPTION
**Pull request by Mte90**
Added support for a new parameter `api_url` for other wakatime host replacements.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -13,10 +13,7 @@ include(KDECMakeSettings)
 
 # Sane flags from Kate project
 add_definitions(-DQT_USE_QSTRINGBUILDER
-                -DQT_NO_CAST_TO_ASCII
-                -DQT_NO_CAST_FROM_ASCII
                 -DQT_STRICT_ITERATORS
-                -DQT_NO_URL_CAST_FROM_STRING
                 -DQT_NO_CAST_FROM_BYTEARRAY
```
Comment (Tatsh) [inline CMakeLists.txt:17]: Why were these removed?

File: wakatimeplugin.cpp [hunk 2]
```diff
@@ -120,7 +120,7 @@ WakaTimeView::WakaTimeView(KTextEditor::MainWindow *mainWindow)
     connect(nam,
             &QNetworkAccessManager::finished,
             this,
-            &WakaTimeView::slotNetworkReplyFinshed);
+            &WakaTimeView::slotNetworkReplyFinished);
```
Comment (Tatsh) [inline wakatimeplugin.cpp:123]: Thanks for the spelling fix.

**Review by Tatsh — COMMENTED** [review]
This looks good. Could you rebase/squash this into a single commit?

File: CMakeLists.txt (hunk 1 above)
Comment (Mte90) [reply CMakeLists.txt:17]: As don't compile with that. I am not so much expert in C++ and Qt but I got some errors when I used QString.

**Comment by Mte90** [comment]
> This looks good. Could you rebase/squash this into a single commit?

I am trying to understand why I am getting a 401 if I use https://wakapi.dev/ and not with the official service.
Also I am not sure how to get debug info as I am not so skilled with C++ and Qt network APIs.

**Comment by Mte90** [comment]
So I updated the code also for Qt 5.15 to avoid some deprecation errors.
There is one that I don't know how to fix:
```
/home/mte90/Desktop/kde/kate-wakatime/wakatimeplugin.cpp: In member function ‘void WakaTimeView::sendQueuedHeartbeats()’:
/home/mte90/Desktop/kde/kate-wakatime/wakatimeplugin.cpp:384:32: warning: ‘QNetworkAccessManager::NetworkAccessibility QNetworkAccessManager::networkAccessible() const’ is deprecated [-Wdeprecated-declarations]
  384 |     if (nam->networkAccessible() != QNetworkAccessManager::Accessible) {
      |                                ^
In file included from /usr/include/x86_64-linux-gnu/qt5/QtNetwork/QNetworkAccessManager:1,
                 from /home/mte90/Desktop/kde/kate-wakatime/wakatimeplugin.cpp:46:
/usr/include/x86_64-linux-gnu/qt5/QtNetwork/qnetworkaccessmanager.h:160:53: note: declared here
  160 |     QT_DEPRECATED_VERSION_5_15 NetworkAccessibility networkAccessible() const;
```

About merge in a single commit you can do in Github the squash to a single commit when you press the green button on the arrow.

**Comment by Tatsh** [comment]
Are you going to push more or do you want me to try this locally? I want to see about removing the changes in CMakeLists.txt and the deprecation issue with QNetworkAccess.

That part is just a warning. It seems Qt project does not want those functions to be used anymore. So I guess a replacement to check if we are offline needs to be put in. In particular, probably just want to check if `apiUrlPath` can be used at all. If it fails a pre-flight then assume we are offline. If you want to implement this, please do so.

**Comment by Mte90** [comment]
I finished working on that PR :-)

File: wakatimeplugin.cpp [hunk 3]
```diff
@@ -435,13 +435,13 @@ void WakaTimeView::sendHeartbeat(const QVariantMap &data,
     QByteArray requestContent = object.toJson();
     static const QString contentType = QLatin1String("application/json");
 
-    static QUrl url(QLatin1String(kWakaTimeViewActionUrl));
+    static QUrl url(QString("%1/v1/actions").arg(apiUrl));
```
Comment (ognarb) [inline wakatimeplugin.cpp:438]:
```suggestion
    static QUrl url(QString(QStringLiteral("%1/v1/actions")).arg(apiUrl));
```

File: wakatimeplugin.cpp [hunk 4]
```diff
@@ -505,8 +507,14 @@ void WakaTimeView::readConfig(void) {
         return;
     }
 
+    QString url = QString("https://wakatime.com/api");
```
Comment (ognarb) [inline wakatimeplugin.cpp:510]:
```suggestion
    QString url = QStringLiteral("https://wakatime.com/api");
```

File: CMakeLists.txt (hunk 1 above)
Comment (ognarb) [reply CMakeLists.txt:17]: You need to use QStringLiteral("mystrin") instead of "mystring"

**Review by Tatsh — APPROVED** [review]

**Comment by Mte90** [comment]
Thanks!
